test/Grpc.AspNetCore.Server.Tests/HttpContextServerCallContextTests.cs

```diff
@@ -529,11 +529,96 @@ public void RequestHeaders_ManyHttpRequestHeaders_HeadersFiltered(string headerN
             Assert.AreEqual(addedToRequestHeaders, headerAdded);
         }
 
+        [Test]
+        public async Task Dispose_LongRunningDeadlineAbort_WaitsUntilDeadlineAbortIsFinished()
+        {
+            // Arrange
+            var cts = new CancellationTokenSource();
+            cts.CancelAfter(TimeSpan.FromSeconds(1));
+
+            var httpContext = new DefaultHttpContext();
+            httpContext.Request.Headers[GrpcProtocolConstants.TimeoutHeader] = "100n";
+            httpContext.Features.Set<IHttpRequestLifetimeFeature>(new TestBlockingHttpRequestLifetimeFeature(cts));
+
+            var testSink = new TestSink();
+            var testLogger = new TestLogger(string.Empty, testSink, true);
+
+            var serverCallContext = CreateServerCallContext(httpContext, testLogger);
+            serverCallContext.Initialize();
+
+            // Wait until we're inside the deadline method and the lock has been taken
+            while (true)
+            {
+                if (testSink.Writes.Any(w => w.EventId.Name == "DeadlineExceeded"))
+                {
+                    break;
+                }
+            }
+
+            // Act
+            var disposeTask = Task.Run(() => serverCallContext.Dispose());
+
+            // Assert
+            if (await Task.WhenAny(disposeTask, Task.Delay(TimeSpan.FromSeconds(0.2))) == disposeTask)
```

Comment:
☹️ not a fan of using an arbitrary timespan to test this.

Reply:
Neither am I but it is tough to unit test this locking.